### PR TITLE
Make struct_union_typet::component_type return a const reference

### PR DESCRIPTION
**Pull request by tautschnig**
There is no need for a copy.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: b7b2eb1a11796074c756621a3c8cb5e7c4da0882).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85277924

**Review by kroening — APPROVED** [review]

File: jbmc/src/java_bytecode/java_string_library_preprocess.cpp [hunk 1]
```diff
@@ -392,7 +392,8 @@ java_string_library_preprocesst::process_equals_function_operands(
 /// \param type: a type containing a "data" component
 /// \param symbol_table: symbol table
 /// \return type of the "data" component
-static typet get_data_type(const typet &type, const symbol_tablet &symbol_table)
+static const typet&
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_string_library_preprocess.cpp]: Suggest also taking the symbol on line 401 by reference, as this currently comes worryingly close to returning a reference to an `irept` (the handle structure, not the `dt` within) allocated on the stack, though I think the final `component_type` call avoids this fate.

Comment (tautschnig) [reply jbmc/src/java_bytecode/java_string_library_preprocess.cpp]: Done.

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 63044d6ccccf6a989ff36783d7165b873d61ed83).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85546191

**Pull request merged**